Feat : 로그아웃 기능

jungjinbeom opened the pull request:
### 로그아웃 
- 로그아웃 API 연동
- 로그아웃 API 호출 커스텀 훅 추가
- LocalStorage 삭제 유틸 함수 추가
- Header 컴포넌트에 로그아웃 함수 추가

jungjinbeom committed:
feat : 로그아웃 기능
- 로그아웃 API 연동
- 로그아웃 커스텀 훅 추가
- 로컬스토리지 삭제 유틸 함수 추가
- 헤더 컴포넌트에 로그아웃 함수 추가

Seung-wan commented on src/components/layout/header/header.tsx:
```diff
@@ -1,8 +1,19 @@
 const Header = () => {
-  const handleLogout = () => {};
+  const navigate = useNavigate();
+  const logoutMutation = useLogoutMutate();
```
요런 느낌으로 표현해주는게 더 괜찮지 않을까요?
```suggestion
  const { mutate:logoutMutate } = useLogoutMutate();
```

Seung-wan replied:
요렇게 리액트 쿼리의 훅 이름을 그대로 따라주는게 좋지 않을까요?
```suggestion
  const logoutMutation = useLogoutMutation();
```

jungjinbeom committed:
chore: header 컴포넌트 import 수정
- header 컴포넌트 import문 각각 역할에 맞게 순서 변경

Seung-wan commented on src/components/layout/header/header.tsx:
```diff
@@ -1,8 +1,19 @@
-  const handleLogout = () => {};
+  const navigate = useNavigate();
+  const logoutMutation = useLogoutMutate();
+  const logout = () => {
+    removeLocalStorage('token');
+    logoutMutation.mutate();
+    navigate(ROUTE_PATH.LOGIN);
+  };
```
로그아웃이 성공해야 그때 로컬스토리지의 콘을 지워주고 navigate를 시켜야하는거 아닐까요?
지금은 로그아웃이 실패해도 토큰이 날라가고 페이지가 이동될 것 같아요.

Seung-wan commented on src/components/layout/header/header.tsx:
```diff
@@ -1,8 +1,19 @@
+  const navigate = useNavigate();
+  const logoutMutation = useLogoutMutate();
+  const logout = () => {
+    removeLocalStorage('token');
```
key값을 이렇게 문자열로 전달할 필요 없도록 localStorage 모듈화를 해보면 어떨까요?
사용하는곳마다 'token'이라는 값을 넣어주는 것은 불필요한 행위일 것 같아요.

jungjinbeom replied:
사용하면서 token이라는 값을 넣어주는게 불필요하다고 생각했는데 token 값을 가져올수 있는 함수나 훅을 따로 개발해야될거 같아요

Seung-wan replied:
class 방식으로 Storage라는 것을 하나 만들고, 인스턴스를 생성할 때 key값을 넣어줄수도 있을 것 같구요!
아니면 클로저를 이용해서도 구현해도 될 것 같네요!

jungjinbeom committed:
chore: header 컴포넌트 함수 개행 추가
- 가독성을 위해 커스터훅과 함수 사이의 개행 추가

Seung-wan commented on src/hooks/useLogoutMutate.ts:
리액트 쿼리의 훅은 `src/hooks/queries`와 같이 계층을 하나 더 만들어서 모아둬도 괜찮을 것 같아요.
리액트 쿼리라는 라이브러리를 활용한 훅이니까요!

Seung-wan commented on src/utils/storage.ts:
```diff
@@ -1,7 +1,13 @@
-export const setLocalStorage = ({ key, value }: { key: string; value: string }) => {
+type SetLocalStorage = Record<'key' | 'value', string>;
+
+export const setLocalStorage = ({ key, value }: SetLocalStorage) => {
```
localStorage를 순수하게 사용할때도 key, value 순서로 그대로 넣다보니, 여기서 객체로 표현되면 오히려 이 모듈을 사용하는 입장에서 더 어색할수도 있겠다는 생각이 드네요!
객체가 아닌 그대로 key, value로 매개변수를 표현해줘도 괜찮을 것 같아요.

jungjinbeom replied:
다른 localStorage에서는 매개변수로 사용하고 있어 통일성을 위해서 매개변수로 변경하는게 맞는거 같아요

jungjinbeom committed:
chore: logout 커스텀 훅 명칭 수정 및 구조 분해 할당 적용
- logout 커스텀 훅 명칭 수정 -> useLogoutMutation
-  useLogoutMutation 훅 사용 시 구조 분해 할당 적용하여 mutate 함수 사용

jungjinbeom committed:
chore : 로그아웃 예외처리
- 로그아웃 API 실패시 예외처리를 위해 try~catch 추가

jungjinbeom committed:
chore: 로그아웃 예외처리 수정
- logout API 호출 후 행동이 끝날 때까지 대기시키기 위해 async await 추가
